register-machines/src/main.rs

```diff
@@ -8,44 +8,114 @@ type State = (Label, HashMap<Register, u128>);
 enum Instruction {
     Add(Register, Label),
     Sub(Register, Label, Label),
-    Halt
+    Halt,
 }
-
 use Instruction::*;
+
+#[allow(dead_code)]
 fn eval_program(program: &[Instruction], state: &State) -> State {
-    unimplemented!();
+    let (mut label, mut regs) = state.clone();
+    while (0..program.len()).contains(&label) {
+        match program[label] {
+            Add(reg, new_label) => {
+                *regs.entry(reg).or_default() += 1;
+                label = new_label;
+            }
+            Sub(reg, new_label, other_label) => {
+                let register = regs.entry(reg).or_default();
+                label = if *register > 0 {
+                    *register -= 1;
+                    new_label
+                } else {
+                    other_label
+                }
+            }
+            Halt => break,
+        };
+    }
+    (label, regs)
 }
 // <<x,y>> = (2^x)*(2y+1)
+#[allow(dead_code)]
 fn encode_pair1(x: u128, y: u128) -> u128 {
-    unimplemented!();
+    (1 << x) * (2 * y + 1)
 }
```

Comment:
The compiler likely does this, but don't put too much faith in rustc just yet:
`#[inline(always)]`